From 76c14b89db6a45f479e4661585c889b0f54817c5 Mon Sep 17 00:00:00 2001
From: Shuming Fan <shumingf@realtek.com>
Date: Thu, 30 Oct 2025 19:30:16 +0800
Subject: [PATCH] ucm2: sof-soundwire: support rt713vb codec

The rt713 is the multi-function codec, so check it by MultiCodec variable.
This patch also supports the rt713VB, which integrates with DMIC function into one interface.

Closes: https://github.com/alsa-project/alsa-ucm-conf/pull/628
Signed-off-by: Shuming Fan <shumingf@realtek.com>
Signed-off-by: Jaroslav Kysela <perex@perex.cz>
Origin: backport, https://github.com/alsa-project/alsa-ucm-conf/commit/76c14b89db6a45f479e4661585c889b0f54817c5
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2129952
---
 ucm2/codecs/rt713-sdca/init.conf              |  8 -----
 ucm2/codecs/rt713/init.conf                   | 23 +++++++++++++++
 .../{rt713-sdca.conf => rt713.conf}           | 29 +++++++++++++++++++
 ucm2/sof-soundwire/sof-soundwire.conf         |  8 ++---
 4 files changed, 56 insertions(+), 12 deletions(-)
 delete mode 100644 ucm2/codecs/rt713-sdca/init.conf
 create mode 100644 ucm2/codecs/rt713/init.conf
 rename ucm2/sof-soundwire/{rt713-sdca.conf => rt713.conf} (58%)

--- a/ucm2/codecs/rt713-sdca/init.conf
+++ /dev/null
@@ -1,8 +0,0 @@
-# RT713-sdca specific volume control settings
-
-BootSequence [
-	cset "name='rt713 FU05 Playback Volume' 87"
-	cset "name='rt713 ADC 23 Mux' 'MIC2'"
-	cset "name='rt713 FU0F Capture Volume' 57"
-	cset "name='rt713 FU0F Capture Switch' 1"
-]
--- /dev/null
+++ b/ucm2/codecs/rt713/init.conf
@@ -0,0 +1,23 @@
+# RT713 specific volume control settings
+
+BootSequence [
+	cset "name='rt713 FU05 Playback Volume' 87"
+	cset "name='rt713 ADC 23 Mux' 'MIC2'"
+	cset "name='rt713 FU0F Capture Volume' 57"
+	cset "name='rt713 FU0F Capture Switch' 1"
+]
+
+If.rt713vb_init {
+	Condition {
+		Type String
+		String1 "rt713"
+		String2 "${var:MultiMicShadow}"
+	}
+	True.BootSequence [
+		# RT713-VB integrated with DMIC
+		cset "name='rt713 FU1E Capture Switch' 1"
+		cset "name='rt713 FU1E Capture Volume' 47"
+		cset "name='rt713 ADC 0A Mux' 'DMIC1'"
+		cset "name='rt713 ADC 0B Mux' 'DMIC2'"
+	]
+}
--- a/ucm2/sof-soundwire/rt713-sdca.conf
+++ /dev/null
@@ -1,40 +0,0 @@
-# Use case Configuration for sof-soundwire card
-
-SectionDevice."Headphones" {
-	Comment	"Headphones"
-
-	EnableSequence [
-		cset "name='Headphone Switch' on"
-	]
-
-	DisableSequence [
-		cset "name='Headphone Switch' off"
-	]
-
-	Value {
-	      PlaybackPriority 200
-	      PlaybackPCM "hw:${CardId}"
-	      JackControl "Headphone Jack"
-	}
-}
-
-SectionDevice."Headset" {
-	Comment "Headset Microphone"
-
-	EnableSequence [
-		cset "name='rt713 FU0F Capture Switch' 1"
-	]
-
-	DisableSequence [
-		cset "name='rt713 FU0F Capture Switch' 0"
-	]
-
-	Value {
-		CapturePriority 200
-		CapturePCM "hw:${CardId},1"
-		JackControl "Headset Mic Jack"
-		CaptureSwitch "rt713 FU0F Capture Switch"
-		CaptureVolume "rt713 FU0F Capture Volume"
-		CaptureMixerElem "rt713 FU0F"
-	}
-}
--- /dev/null
+++ b/ucm2/sof-soundwire/rt713.conf
@@ -0,0 +1,69 @@
+# Use case Configuration for sof-soundwire card
+
+SectionDevice."Headphones" {
+	Comment	"Headphones"
+
+	EnableSequence [
+		cset "name='Headphone Switch' on"
+	]
+
+	DisableSequence [
+		cset "name='Headphone Switch' off"
+	]
+
+	Value {
+	      PlaybackPriority 200
+	      PlaybackPCM "hw:${CardId}"
+	      JackControl "Headphone Jack"
+	}
+}
+
+SectionDevice."Headset" {
+	Comment "Headset Microphone"
+
+	EnableSequence [
+		cset "name='rt713 FU0F Capture Switch' 1"
+	]
+
+	DisableSequence [
+		cset "name='rt713 FU0F Capture Switch' 0"
+	]
+
+	Value {
+		CapturePriority 200
+		CapturePCM "hw:${CardId},1"
+		JackControl "Headset Mic Jack"
+		CaptureSwitch "rt713 FU0F Capture Switch"
+		CaptureVolume "rt713 FU0F Capture Volume"
+		CaptureMixerElem "rt713 FU0F"
+	}
+}
+
+If.codecmic {
+	Condition {
+		Type String
+		String1 "rt713"
+		String2 "${var:MultiMicShadow}"
+	}
+	True {
+		SectionDevice."Mic" {
+			Comment "SoundWire Microphones"
+
+			EnableSequence [
+				cset "name='rt713 FU1E Capture Switch' 1"
+			]
+
+			DisableSequence [
+				cset "name='rt713 FU1E Capture Switch' 0"
+			]
+
+			Value {
+				CapturePriority 100
+				CapturePCM "hw:${CardId},4"
+				CaptureSwitch "rt713 FU1E Capture Switch"
+				CaptureVolume "rt713 FU1E Capture Volume"
+				CaptureMixerElem "rt713 FU1E"
+			}
+		}
+	}
+}
--- a/ucm2/sof-soundwire/sof-soundwire.conf
+++ b/ucm2/sof-soundwire/sof-soundwire.conf
@@ -45,7 +45,7 @@
 		String "${CardComponents}"
 	}
 	MultiCodec {
-		Regex "(rt712|rt721|rt722(-sdca)?)"
+		Regex "(rt712|rt713|rt721|rt722(-sdca)?)"
 		String "${var:SpeakerCodec1} ${var:HeadsetCodec1} ${var:MicCodec1}"
 	}
 }
@@ -88,9 +88,9 @@
 
 If.multi_headset {
 	Condition {
-		Type RegexMatch
-		Regex "${var:HeadsetCodec1}"
-		String "${var:MultiCodec1}"
+		Type String
+		Haystack "${var:HeadsetCodec1}"
+		Needle "${var:MultiCodec1}"
 	}
 	True {
 		Define.HeadsetCodec1 ""
